[#8] 권한 부여 및 검사

daakludens opened the pull request:
관리자, 게임 공급 회사, 이용자(유저, 게이머)에 따른 게임 관리에 대한 권한 부여 기능 추가

daakludens committed:
[#8] feat: 사용자 레벨을 관리하기 위한 AuthLevel enum 생성

daakludens committed:
[#8] feat: AuthLevel 사용를 적용한 GameController 추가

daakludens committed:
[#8] feat: httpSession에 AuthLevel 권한 담는 기능 추가

daakludens committed:
[#8] refactor: mybatis underscore 컬럼명 자동 camel case 되도록 설정, UserMapper alias 제거

daakludens committed:
[#8] refactor: 세션에 삽입할 AuthLevel의 attribute 이름을 ROLE로 일치시킴

f-lab-dev commented on src/main/java/com/ludensdomain/controller/GameController.java:
```diff
@@ -0,0 +1,58 @@
+    AuthLevel authLevel;
+
+    @GetMapping("/{gameId}")
+    public ResponseEntity<Void> selectGame(@PathVariable String gameId) {
```
이슈는 최대한 잘게 쪼개서 진행하시는게 낫습니다.
게임 기능과 권한관리는 아래 같이 세부적인 이슈로 나눌 수 있겠네요

- 게임 불러오기
- 수정
- 삭제
- 등록
- 권한나누기

등등 많이 나눌 수 있을 것 같은데 각 이슈 하나하나마다 깊게 고민할만한 포인트가 꽤 많기때문에 여러 이슈를 늘어놓고 고민하기보다는 한 이슈씩 집중하는것이 좋습니다.

지금이라도 이슈를 잘게 나눠서 단순 CRUD가 아닌 `각 하나하나의 기능을 더 잘 만들려면 어떻게 해야할지`에 대해 고민해보시면 좋을 것 같아요

daakludens replied:
이제 보니 단순 값 넘기기로 진행하고 있었네요. 제가 생각이 짧았습니다. 
그러면 해당 브랜치는 제일 많이 진행한 권한 나누기 및 검사 기능으로 진행하고 이슈도 수정하겠습니다.

또한 이슈를 세부화 한 다음 comment에 올려서 리뷰 받을 수 있을까요?

f-lab-dev replied:
이슈 나눈것에 대한 리뷰를 받는다는 말씀이시죠? 나누시고 슬랙에 말씀주시면 될 것 같네요~

f-lab-dev commented on src/main/java/com/ludensdomain/controller/GameController.java:
```diff
@@ -0,0 +1,58 @@
+    @PostMapping("/{gameId}")
+    public ResponseEntity<Void> insertNewGame(@PathVariable String gameId) {
+        if (authLevel == AuthLevel.ADMIN) {
+            return RESPONSE_OK;
```
이렇게 권한 검사하는 코드도 분리가 가능할 것 같은데 이것도 이슈로 등록해서 작업해주세요~

daakludens replied:
네, 수정하겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/service/SessionLoginService.java:
```diff
@@ -19,6 +21,21 @@ public boolean isLoginUser(long id) {
+    public void authLevelByRole(UserDto user) {
+        String role = user.getRole();
+
+        if (role.equals("1")) {
```
혹시 모를 NPE를 막기위해 상수를 앞에다 두는게 좋습니다. `"1".equals(role)`

daakludens replied:
아, 그렇군요. 
질문 있습니다. 상수를 앞에 두는 것도 좋지만 role의 null check를 미리 한 번 해주는 것도 안전한 방법이라는 의견들도 있었습니다. equals( ) 전에 role의 null check를 하는 건 어떤가요? 아니면 exception 가능성이 있으니 throws를 추가하는 건요?

f-lab-dev replied:
네 그렇게 제대로 된 값이 들어왔는지 유효성검사를 해주는것도 방법입니다~

f-lab-dev commented on src/main/java/com/ludensdomain/service/SessionLoginService.java:
```diff
@@ -19,6 +21,21 @@ public boolean isLoginUser(long id) {
+            httpSession.setAttribute("ROLE", AuthLevel.ADMIN);
+        }
+        if (role.equals("2")) {
+            httpSession.setAttribute("ROLE", AuthLevel.COMPANY);
```
`"ROLE"`은 세션의 키니까 상수로 관리하면 유지보수성이 더 좋아질 것 같네요

daakludens replied:
네, 그럼 필드에 public static final String ROLE을 추가해 상수화 시키겠습니다.

daakludens committed:
##### [#8] : refactor
- 세션 role 권한 부여 equals 메서드에 상수를 앞에 둬 NPE 방지
- role null check로 유효성 검사

daakludens committed:
##### [#8] : refactor
- 세션 key 상수화

daakludens committed:
[#8]feat :
중복적 권한 체크의 성능 개선을 위한 aop 적용
- LoginCheck와 LoginCheckAspect 추가
- GameController에 생성한 LoginCheck 어노테이션 적용

daakludens committed:
[#8]refactor :
- 오타 수정

f-lab-dev commented on src/main/java/com/ludensdomain/service/SessionLoginService.java:
```diff
@@ -25,14 +25,16 @@ public boolean isLoginUser(long id) {
-        if (role.equals("3")) {
-            httpSession.setAttribute("ROLE", AuthLevel.USER);
+        if (role != null) {
+            if ("1".equals(role)) {
```
이거는 switch-case 문법을 쓰는게 더 깔끔하지 않을까요?

daakludens replied:
네, 너무 if 문이 많은 것 같네요. switch-case 문법으로 변경하겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -0,0 +1,28 @@
+    public void loginCheck() {
+        AuthLevel authLevel = (AuthLevel) httpSession.getAttribute("ROLE_COMPANY");
+
+        if (authLevel != AuthLevel.COMPANY) {
```
이거는 현재 `COMPANY` role만 검사하게 되어있는데 추후에 확장성을 위해 여러 role을 검사할 수 있으면 좋겠습니다~

daakludens replied:
네, COMPANY 뿐만 아니라 USER와 ADMIN도 권한 체크가 되도록 설정하겠습니다.

daakludens committed:
[#8]refactor :
- SessionLoginService에서 권한 분리/적용 조건을 switch-case 문법으로 변경

daakludens committed:
[#8] feat :
3가지 권한 종류를 전부 체크할 수 있도록 어노테이션 추가
- admin과 user 권한 체크를 위한 AdminCheck와 UserCheck 추가
- LoginCheckAspect에 어노테이션 추가
- GameController에 게임 조회 기능에 user 권한 체크 추가

f-lab-dev commented on src/main/java/com/ludensdomain/service/SessionLoginService.java:
```diff
@@ -29,14 +29,18 @@ public void authLevelByRole(UserDto user) {
-            }
-            if ("3".equals(role)) {
-                httpSession.setAttribute(ROLE_USER, AuthLevel.USER);
+            switch (role) {
```
switch-case를 쓰기 때문에 위의 `if (role != null) `는 지워줘도 되지 않을까요?

daakludens replied:
null인 경우엔 default로 넘어가면서 처리가 되겠네요. 수정하겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/service/SessionLoginService.java:
```diff
@@ -29,14 +29,18 @@ public void authLevelByRole(UserDto user) {
+                case "2" :
+                    httpSession.setAttribute(ROLE_COMPANY, AuthLevel.COMPANY);
+                    break;
+                case "3" :
```
이 값은 하드코딩 되어있는데 상수나 enum을 쓰는게 좋아보입니다

daakludens replied:
하드코딩을 줄이기 위해서 AuthLevel 클래스에서 switch 문으로 권한 부여하는 방식으로 변경했습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/aop/AdminCheck.java:
```diff
@@ -7,6 +7,6 @@
 @Retention(RetentionPolicy.RUNTIME)
 @Target(ElementType.METHOD)
-public @interface LoginCheck {
+public @interface AdminCheck {
```
굳이 권한을 각각 어노테이션으로 따로 나눌 필요가 있을까요? role을 이 어노테이션의 인자로 받으면 어노테이션 하나만으로 표현이 가능해보입니다~

daakludens replied:
GameController에서 @path variable으로 role을 받아서 @Before 어노테이션 인자로 넘기면 될 것 같습니다.

f-lab-dev replied:
굳이 그럴 필요가 있을까요? 어노테이션에도 인자를 설정할 수 있습니다~

daakludens committed:
[#8] refactor :
- SessionLoginService의 switch 바깥의 null check 삭제

daakludens committed:
[#8] refactor :
- AuthLevel 클래스에 권한 식별 로직 이동
- SessionLoginService 클래스에 권한 식별 로직 삭제
- aop 사용을 위한 어노테이션 추가

f-lab-dev commented on src/main/java/com/ludensdomain/controller/UserController.java:
```diff
@@ -45,7 +48,7 @@
             result = RESPONSE_BAD_REQUEST;
         } else {
-            loginService.authLevelByRole(user);
+            httpSession.setAttribute("ROLE", authLevelByRole(user.getRole()));
```
여기도 세션 키로 사용하는건데 유지보수성을 위해 상수를 쓰시는게 나아보이네요

daakludens replied:
네, 세션키인 "ROLE"을 상수로 선언해 변경했습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/domain/AuthLevel.java:
```diff
@@ -1,5 +1,21 @@
 public enum AuthLevel {
-    ADMIN, COMPANY, USER
+
+    ADMIN("1"), COMPANY("2"), USER("3");
```
꼭 숫자로 표현할 필요가 있을까요?

daakludens replied:
네, 숫자 제외했습니다.

daakludens committed:
[#8] refactor :
- LoginCheckAspect 포인트컷 하나로 압축
- GameController 어노테이션 변경

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -16,31 +15,50 @@
-        AuthLevel authLevel = (AuthLevel) httpSession.getAttribute("ROLE_ADMIN");
+    @Before("@annotation(com.ludensdomain.aop.UserCheck) && @annotation(UserCheck)")
+    public void userCheck(AuthLevel authLevel) throws Exception {
+        if (authLevel == AuthLevel.ADMIN) {
```
switch-case로 처리하면 조금 더 깔끔할 것 같습니다~

daakludens replied:
네, switch-case로 시정하겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -16,31 +15,50 @@
+
+    private long getCurrentUser() {
+
+        return Long.parseLong(httpSession.getId());
```
`httpSession.getId()` 메소드에선 어떤 값이 리턴되나요? DB에 있는 그 유저의 ID 값이 맞을까요?

daakludens replied:
`httpSession.getId()` 는 제가 가지려 했던 세션 내의 유저 아이디가 아닌 세션 아이디, 즉 JSESSIONID를 반환하게 됩니다. 제가 예상했던 값이 아니네요. 수정하겠습니다.

f-lab-dev replied:
네 모든 코드는 실제로 잘 동작하는지 돌려보고 push해주세요~

daakludens committed:
[#8] refactor :
* @interface 클래스 3개를 LoginCheck 하나로 통합
* UserCheck 클래스를 LoginCheck 클래스로 이름 변경
* LoginCheckAspect에서 if-else문을 switch문으로 변경
* GameCOntroller에 변경된 @interface 반영

daakludens commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -17,47 +16,49 @@
-            companyCheck();
+    @Before("@annotation(com.ludensdomain.aop.LoginCheck)")
+    public void userCheck(LoginCheck.AuthLevel target) throws Exception {
+        switch (target) {
```
멘토님, LoginCheckAspect 클래스의 19번째 줄의 @Before에 계속 unbound pointcut parameter 에러가 표시됩니다. 분명히 LoginCheck를 찾을 수 있도록 포인트컷을 괄호 안에 명시해줬는데도 에러가 나는 이유를 모르겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -17,47 +16,49 @@
+    @Before("@annotation(com.ludensdomain.aop.LoginCheck)")
+    public void userCheck(LoginCheck.AuthLevel target) throws Exception {
+        switch (target) {
+            case ADMIN: adminCheck();
```
break까지 포함하여 두 줄이므로 여기도 개행 한번 해주시면 좋겠네요~

daakludens replied:
네, 메서드 호출은 줄 나눔으로 아래 줄로 내렸습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -17,47 +16,49 @@
+            case USER: userCheck();
+                break;
+            default:
+                throw new IllegalStateException("Unexpected value: " + target);
```
이거는 잘못된 값을 넘긴거니까 `IllegalArgumentException`이 낫지 않을까요?

daakludens replied:
자바 문서에서 IllegalStateException은 `메서드의 호출 타이밍이 부적절`할 때 쓰이는 예외처리고, IllegalArgumentException은 `부적잘한 인자`가 들어왔을 때 쓰이는 예외 처리라고 소개합니다. 
그러므로 target에 잘못 들어온 값을 거르는 예외 처리로는 `IllegalArgumentException`이 정확한 예외입니다.

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -17,47 +16,49 @@
     }
 
     private long getCurrentUser() {
+        Object obj = (httpSession.getAttribute("id"));
```
여기서 괄호는 왜 넣어주신걸까요?

daakludens replied:
데이터 타입으로 씨름하다가 깜빡했습니다. 수정하겠습니다.

daakludens committed:
[#8] refactor :
1. LoginCheckAspect의 switch문 case와 break 부분 개행
2. getCurrentUser 메서드에 불필요한 괄호 삭제
3. 예외처리명을 IllegalArgumentException으로 변경

daakludens committed:
[#8] fix: aop unbound pointcut parameter 에러 수정

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -41,27 +42,27 @@ private long getCurrentUser() {
     private void adminCheck() throws Exception {
         long id = getCurrentUser();
-        LoginCheck.AuthLevel role = userService.findUserById(id).getRole();
+        AuthLevel role = userService.findUserById(id).getRole();
```
role은 세션에 저장하고 가져와도 될 것 같네요~ 이렇게 하면 요청을 보낼때마다 mysql에 부하를 주게될 것 같습니다

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -41,27 +42,27 @@ private long getCurrentUser() {
 
-        if (!(role == LoginCheck.AuthLevel.ADMIN)) {
+        if (!(role == AuthLevel.ADMIN)) {
             throw new Exception();
```
이렇게 빈 예외를 던지게 되면 디버깅하는 입장에서는 예외를 받아도 이 예외가 왜 발생했는지 파악하기 힘듭니다~ 자세하게 메세지와 타입을 정해서 던져주세요

daakludens committed:
[#8] refactor: httpSession에 ROLE을 보관해 바로 값을 가져오도록 변경

daakludens committed:
[#8] refactor: 예외 발생 시 정확한 예외 원인을 파악하기 위해 빈 예외에서 IllegalArgumentException으로 변경

f-lab-dev commented on src/main/java/com/ludensdomain/controller/UserController.java:
```diff
@@ -40,6 +41,8 @@
             result = RESPONSE_BAD_REQUEST;
         } else {
+            AuthLevel role = userService.findUserById(id).getRole();
+            httpSession.setAttribute(ROLE, role);
```
이 전체코드를 보니 `유저가 있는지 확인하고 있다면 세션에 값을 넣는다`는 비즈니스 로직입니다.
그러면 이거는 Service에 들어가는게 맞지 않을까요?

daakludens replied:
질문 있습니다. 해당 두 줄에서 user 관련 로직도 들어가고 세션 관련 로직도 들어가는데 어느 쪽 service에 코드를 작성해야 하나요? 
제 생각에는 두 로직을 분리할 수 있기 때문에 userService에 user의 role을 추출한 다음 해당 메서드에 SessionLoginService의 메서드를 호출해 role을 삽입하려고 합니다. 
그리고 지금 보니까 이미 위에서 유저가 있는지 확인을 하고 있어서 findUserById는 중복적인 것 같네요.

f-lab-dev replied:
`LoginService`가 `UserService`에 의존하고있으니 `LoginService`에 넣는게 깔끔하지 않을까요?

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -31,31 +33,31 @@ public void loginCheck(LoginCheck loginCheck) throws Exception {
     }
 
-    private void adminCheck() throws Exception {
+    private void adminCheck() {
```
그리고 이 메소드는 굳이 분리할 필요는 없을 것 같습니다~ 단순히 role을 검사하고 예외를 던져주는 역할인데 위의 메소드에서 이미 role을 검사하고있죠. 그래서 중복 검사인 것 같네요~

daakludens replied:
컨트롤러에서 받아 오는 authLevel과 httpSession에 저장된 role이 같은지 확인하기 위해서는 각 role의 메서드를 만들어서 확인하는 게 필요한 거 아닌가요?

f-lab-dev replied:
네 맞습니다~ 다만 그것을 검사하는것은 `loginCheck`에서만 해줘도 충분할 것 같아요.
switch-case로 나누지 않고 그냥 `컨트롤러에서 받아 오는 authLevel과 httpSession에 저장된 role이 같은지`만 검사해주면 되지 않을까요?

daakludens replied:
아 어차피 비교만 하는 거라 굳이 권한에 따라 메서드를 분리할 필요가 없겠네요. 네, 반영하겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/aop/LoginCheckAspect.java:
```diff
@@ -31,31 +33,31 @@ public void loginCheck(LoginCheck loginCheck) throws Exception {
 
         if (!(role == AuthLevel.ADMIN)) {
-            throw new Exception();
+            throw new IllegalArgumentException(role + " has no access right");
```
`잘못된 접근`이라는 것을 나타내는 예외가 있지 않을까요?

daakludens replied:
잘못된 접근용 예외로는 IllegalAccessException이 있습니다. 해당 예외를 사용하겠습니다.

daakludens committed:
[#8] refactor: 권한 체크 기능 간소화

daakludens committed:
[#8] refactor: 유저 로그인 시 권한을 세션에 넣어주는 기능을 SessionLoginService로 이동

f-lab-dev commented on src/main/java/com/ludensdomain/service/SessionLoginService.java:
```diff
@@ -10,6 +11,13 @@
+    @Override
+    public void insertRole(AuthLevel role) {
+
+        httpSession.setAttribute(ROLE, role);
```
로그인시에 로그인한 유저의 id도 세션에 넣어주는것도 이 클래스에 있어야하지 않을까요?

daakludens replied:
아 id가 세션에 저장되면 매번 db에서 확인할 필요 없이 바로 확인 가능하니 더 유용할 것 같습니다. 수정하겠습니다.

daakludens committed:
[#8] feat. 유저 로그인 시 유저의 아이디와 권한을 세션에 저장

f-lab-dev commented on src/main/java/com/ludensdomain/service/SessionLoginService.java:
```diff
@@ -11,11 +11,12 @@
     @Override
-    public void insertRole(AuthLevel role) {
-
+    public void insertIdAndRole(long id, AuthLevel role) {
```
이 행위는 결국 로그인을 뜻하는 것 같습니다. `loginService`라는 이름에 맞게 `login`으로 메소드 이름을 지어주는게 좋지 않을까요?

daakludens replied:
메서드명이 단순히 매커니즘만 표시하는 게 아니라 기능을 단축적으로 설명해야 하기에 변경하는 게 맞을 것 같습니다.

daakludens committed:
[#8] refactor: 아이디와 권한을 넣어주는 메서드명을 login으로 변경
